File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/Proxy/FileProxyListProvider.java
package ru.kmorozov.gbd.core.logic.Proxy;

import org.apache.commons.io.IOUtils;
import ru.kmorozov.gbd.core.config.GBDOptions;

import java.io.*;
import java.util.HashSet;

/**
 * Created by sbt-morozov-kv on 02.12.2015.
 */
public class FileProxyListProvider extends AbstractProxyListProvider {

    FileProxyListProvider() {
        buildList();
    }

    private void buildList() {
        final String proxyListFileName = GBDOptions.getProxyListFile();
        if (null == proxyListFileName || proxyListFileName.isEmpty()) return;

        final File proxyListFile = new File(proxyListFileName);
        if (!proxyListFile.exists() && !proxyListFile.canRead()) return;

        try (InputStream is = new FileInputStream(proxyListFile)) {
            this.proxyItems = new HashSet(IOUtils.readLines(is, "UTF-8"));
        } catch (final IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void updateProxyList() {
        final String proxyListFileName = GBDOptions.getProxyListFile();
        if (null == proxyListFileName || proxyListFileName.isEmpty()) return;

        final File proxyListFile = new File(proxyListFileName);
        if (!proxyListFile.exists() && !proxyListFile.canRead()) return;

        try (OutputStream os = new FileOutputStream(proxyListFile, false)) {
            for (final HttpHostExt proxy : proxyList)
                if (!proxy.isLocal() && proxy.isAvailable()) IOUtils.write(String.format("%s %s%n", proxy.getHost().getHostName(), proxy.getHost().getPort()), os, "UTF-8");
        } catch (final IOException e) {
            e.printStackTrace();
        }
    }
}

File: ru.kmorozov.gbd.core/src/main/java/ru/kmorozov/gbd/core/logic/proxy/FileProxyListProvider.kt
package ru.kmorozov.gbd.core.logic.proxy

import com.google.common.base.Strings
import ru.kmorozov.gbd.core.config.GBDOptions
import java.io.File
import java.nio.file.Files
import java.util.stream.Collectors

/**
 * Created by sbt-morozov-kv on 02.12.2015.
 */
class FileProxyListProvider internal constructor() : AbstractProxyListProvider() {

    override fun findCandidates() {
        val proxyListFileName = GBDOptions.proxyListFile
        if (Strings.isNullOrEmpty(proxyListFileName)) {
            val proxyListFile = File(proxyListFileName)
            if (proxyListFile.exists() && proxyListFile.canRead())
                this.proxyItems = Files.lines(proxyListFile.toPath()).map { getInetAddress(it) }.collect(Collectors.toSet())
        }
    }

    override fun updateProxyList() {
        val proxyListFileName = GBDOptions.proxyListFile
        if (Strings.isNullOrEmpty(proxyListFileName)) return

        val proxyListFile = File(proxyListFileName)
        val proxyListPath = proxyListFile.toPath()
        if (!proxyListFile.exists() && !proxyListFile.canRead()) return

        Files.write(proxyListPath,
                proxyList.filter { !it.isLocal && it.isAvailable }.map { String.format("%s %s%n", it.host.hostName, it.host.port) })
    }
}
